amqp_serde/src/types.rs

```diff
@@ -415,37 +415,63 @@ impl fmt::Display for FieldValue {
 pub type FieldName = ShortStr;
 /// AMQP field table type.
 #[derive(Serialize, Deserialize, PartialEq, Debug, Clone, Default)]
-pub struct FieldTable(HashMap<FieldName, FieldValue>);
+pub struct FieldTable(LongUint, HashMap<FieldName, FieldValue>);
 
 impl FieldTable {
     pub fn new() -> Self {
-        Self(HashMap::new())
+        Self(0, HashMap::new())
     }
     pub fn insert(&mut self, k: FieldName, v: FieldValue) -> Option<FieldValue> {
-        self.0.insert(k, v)
+        match LongUint::try_from(size_of_val(&k.0) + k.0 as usize + FieldValue::TAG_SIZE + v.len()) {
+            Ok(len) => {
+                self.0 += len;
+                self.1.insert(k, v)
+            },
+            Err(e) => {
+                println!("Error: {}", &e);
```

Comment:
Do you have a better solution for this?

Reply:
@jdelgadoalfonso I prefer your previous solution without changing the FieldTable implementation to avoid complicating implementation.
Can you explain why need to change it?

I guess you are making the changes in order to support public api of `len`?  
I think user do not need to know the internal bytes length, so I prefer to not make the `len`  as public api. And we can allow missing `is_empty` api to supress clippy warning.

Reply:
@gftea, watching the other types that are variable length ([LongStr](https://github.com/gftea/amqprs/blob/main/amqp_serde/src/types.rs#L107), [ByteArray](https://github.com/gftea/amqprs/blob/main/amqp_serde/src/types.rs#L172) and [FieldArray](https://github.com/gftea/amqprs/blob/main/amqp_serde/src/types.rs#L196)), all of them are tuples where the first element contains the data length in bytes.

IMHO, this change makes sense in order to follow the same pattern for all the variable data types and it also makes easier the serde.

The only added logic is to keep tracking length when insert or remove entries from the HashMap. We can simplify the error handler making the supposition that usize -> u32 overflow is quite unlikely.

what do you think?

Reply:
I think less code is best code, so I am a bit biased to have less code if unnecessary, because we know only table type will map to map type , use this knowledge can reduce code, and when adding more code, we will need add more tests to keep coverage.

Reply:
I would like to understand why you didn't implement FieldTable in the same way that LongStr, ByteArray and FieldArray. These last have the length as a field and they also implement TryFrom to convert types and they do not implement an interface to modify them such as: push, pop, ...

I would prefer to implement all of them in the same way, but it is up to you, if you don't change your mind I will implement the len function to calculate byte-length for FieldMap type.

Reply:
First of all, from your PR, you give me some new ideas of design, I do not want to hold the mindset that I own the codes, instead I appreciate the discussion here on design choice, so hope you do not feel I am trying to hold the authority of solution.

In general, it is good to avoid additional length field, but here are my thinking
1. LongStr and ShortStr are both string, but have different length range, one is u32, another is u8, so I add a length field to distinguish them. Do you have a better way?
2. ByteArray and FieldArray are both sequence type, I think with your solution, we can remove the field length for them also. What do you think?
3. The FieldTable is the only type in AMQP that need a `map` type, and vice versa, so based on the knowledge, we do not need to add additional length field.
4. You are right, we should instead remove the push/pop interface, and impl TryFrom, then, user can modify the map before convert to FieldTable.

Are you willing to make such changes in this PR? Or you prefer a new PR?

Reply:
@jdelgadoalfonso  please let me know what you think. Or we can have a voice call on discord channel for discussion, maybe more efficient

Reply:
Sorry for the delayed answer, it was a very busy week.

I appreciate you are open to contributors ideas. I just was trying to understand how you designed the code in order to make the PR aligned to the project.

I would like to contribute to the project.

Regarding to the modification of the FieldTable interface, I suggest to create another PR and make this one dependent from the last one, because the current PR aims to solve an issue deserializing FieldArray which is not directly related to FieldTable interface, however, due to the nature of Rust to keep all use cases implemented, we need to figure out first how to solve FieldTable's length which it is used for the case FieldArray of FieldTables.

To summarize:
   - Create a new PR to:
       1. Remove insert/remove functions.
       2. Make impl TryFrom.
       3. Remove length attribute from ByteArray and FieldArray and impl len function instead.
       4. Modify/Add tests.
   - This PR: (When the above PR be merged)
       1. Use the len function to calculate FieldArray len.
       2. Add tests.

Reply:
Such a good proposal, I agree with that

Reply:
close it, use #70 to track remaining work
 @jdelgadoalfonso thanks a lot for contribution